vm/src/vm.rs

```diff
@@ -328,39 +329,99 @@ impl VirtualMachine {
         PyObject::new(PyModule {}, self.ctx.types.module_type.clone(), Some(dict))
     }
 
+    /// Similar to PyErr_NormalizeException in CPython
+    pub fn normalize_exception(
+        &self,
+        exc_type: PyObjectRef,
+        exc_val: PyObjectRef,
+        exc_tb: PyObjectRef,
+    ) -> PyResult<PyBaseExceptionRef> {
+        let args = || {
+            if self.is_none(&exc_val) {
+                vec![]
+            } else {
+                match_class!(match exc_val.clone() {
+                    tup @ PyTuple => tup.elements.clone(),
+                    exc @ PyBaseException => exc.args().elements.clone(),
+                    obj => vec![obj],
+                })
+            }
+        };
+        let exc_type = if let Ok(exc) = PyBaseExceptionRef::try_from_object(self, exc_type.clone())
```

Comment:
I would not create this `exc_type` variable, but do the condition checking in the code where the exc_type is used below. In fact it might make sense to create another helper function `figure_out_exception` which returns a `PyResult`, so we can do:

```rust
let exc = figure_out_exception(exc_type, exc_value)?;  // figure out a better name for this function ;)
```

Reply:
But that function *is* basically `normalize_exception`. I've reworked the code; I think it's a lot cleaner and easier to follow now.